### PR TITLE
Exploration - Feature Tree / extract helper methods for wrapping tree nodes

### PR DESCRIPTION
**Pull request by juliankrispel**
### Context

This PR is part of a series of PR's that will be exploring **tree data block support** in Draft.

**Extract tree node wrapper methods**

This PR moves the logic for wrapping tree nodes into a separate helper file and replaces a duplicate wrapping effort with said helper (inside `DraftEditorContentsExperimental`)

***

**Note:**
This is unstable and not part of the public API and should not be used by
production systems.

- [x] extracted applyWrapperElementToSiblings
- [x] extracted shouldNotAddWrapperElement
- [x] replace custom block rendering in DraftEditorContentsExperimental with applyWrapperElementToSiblings
- [x] move all logic related to wrapping tree nodes into `wrapBlockNodes`
- [x] add snapshots for wrapBlockNodes

### CODE REVIEW

File: src/component/contents/exploration/DraftEditorBlockNode.react.js [hunk 1]
```diff
@@ -299,43 +249,39 @@ class DraftEditorBlockNode extends React.Component<Props> {
...
+          wrapperTemplate,
           block: child,
         };
 
-        acc.push(
-          React.createElement(
+        acc.push({
+          wrapperTemplate,
+          block: child,
+          element: React.createElement(
```
Comment (juliankrispel) [inline src/component/contents/exploration/DraftEditorBlockNode.react.js]: Before I can give the block nodes to `wrapBlockNodes` I'm mapping it into an array that has this shape `{ wrapperTemplate, block, element }`. Not super confident about this design, lmk if you have a better idea.

File: src/component/contents/exploration/DraftEditorBlockNode.react.js [hunk 2]
```diff
@@ -299,43 +249,39 @@ class DraftEditorBlockNode extends React.Component<Props> {
...
             elementProps,
             <Component {...childProps} />,
           ),
-        );
-
-        if (
-          !wrapperTemplate ||
-          shouldNotAddWrapperElement(child, contentState)
-        ) {
-          return acc;
```
Comment (juliankrispel) [inline src/component/contents/exploration/DraftEditorBlockNode.react.js]: this is now inside `wrapBlockNodes`

File: src/component/contents/exploration/DraftEditorBlockNode.react.js [hunk 3]
```diff
@@ -299,43 +249,39 @@ class DraftEditorBlockNode extends React.Component<Props> {
...
-
+        });
         // if we are here it means we are the last block
         // that has a wrapperTemplate so we should wrap itself
         // and all other previous siblings that share the same wrapper
-        applyWrapperElementToSiblings(wrapperTemplate, Element, acc);
-
         return acc;
       }, []);
+      children = wrapBlockNodes(children, contentState);
```
Comment (juliankrispel) [inline src/component/contents/exploration/DraftEditorBlockNode.react.js]: in order for `shouldNotAddWrapperElement` to work we need contentState I'm afraid...

File: src/component/contents/exploration/DraftEditorContentsExperimental.react.js [hunk 4]
```diff
@@ -146,45 +147,21 @@ class DraftEditorContentsExperimental extends React.Component<Props> {
...
+
       const wrapperTemplate = configForType.wrapper;
+
       processedBlocks.push({
-        block: <DraftEditorBlockNode key={blockKey} {...blockProps} />,
         wrapperTemplate,
-        key: blockKey,
-        offsetKey: DraftOffsetKey.encode(blockKey, 0, 0),
+        block: nodeBlock,
+        element: <DraftEditorBlockNode key={blockKey} {...blockProps} />,
```
Comment (juliankrispel) [inline src/component/contents/exploration/DraftEditorContentsExperimental.react.js]: changed this into the shape the helper function is expecting...

File: src/component/contents/exploration/DraftEditorBlockNode.react.js (hunk 1 above)
Comment (mitermayer) [reply src/component/contents/exploration/DraftEditorBlockNode.react.js]: That sounds reasonable!

File: src/component/contents/exploration/DraftEditorBlockNode.react.js (hunk 3 above)
Comment (mitermayer) [reply src/component/contents/exploration/DraftEditorBlockNode.react.js]: Passing the contentState its ok, now with the new data design thats how we can get the linked blocks

File: src/component/contents/exploration/DraftEditorContentsExperimental.react.js [hunk 5]
```diff
@@ -146,45 +148,21 @@ class DraftEditorContentsExperimental extends React.Component<Props> {
...
       });
 
       const nextBlockKey = nodeBlock.getNextSiblingKey();
       nodeBlock = nextBlockKey ? content.getBlockForKey(nextBlockKey) : null;
     }
 
     // Group contiguous runs of blocks that have the same wrapperTemplate
-    const outputBlocks = [];
-    for (let ii = 0; ii < processedBlocks.length; ) {
```
Comment (mitermayer) [inline src/component/contents/exploration/DraftEditorContentsExperimental.react.js]: So happy with this deleted code! wohooo

File: src/component/utils/exploration/__tests__/wrapBlockNodes.js [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+    const configForType =
+      blockRenderMap.get(block.getType()) || blockRenderMap.get('unstyled');
+
+    const wrapperTemplate = configForType.wrapper;
+    const key = block.getKey();
+    const editorState = EditorState.createWithContent(contentState);
+    const offsetKey = DraftOffsetKey.encode(key, 0, 0);
+
+    const childProps = {
+      blockRenderMap: DefaultDraftBlockRenderMap,
```
Comment (mitermayer) [inline src/component/utils/exploration/__tests__/wrapBlockNodes.js]: We could move this outside of this function and maybe called it something like `DEFAULT_CHILD_PROPS` and inside here we could simply do a:

`
const childProps = {
   ...DEFAULT_CHILD_PROPS,
   ...props
}
`

**Comment by juliankrispel** [comment]
@mitermayer will address the nits. Ping me if and when you can think of more stuff to pass on to me 👍

**Comment by juliankrispel** [comment]
@mitermayer @flarnie all adressed now btw :+1:

File: src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap [hunk 7]
```diff
@@ -0,0 +1,799 @@
...
+      data-offset-key="B-0-0"
+    >
+      <span
+        data-offset-key="B-0-0"
+        style={Object {}}
+      >
+        <span
+          data-text="true"
+        >
+          fist list item
```
Comment (flarnie) [inline src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: Something seems odd to me about this snapshot - do we expect to have two nested 'li' items, each with the text 'first list item'?

Also nit about misspelling of 'first' :)

Comment (flarnie) [reply src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: Yea - looking at this again, I still think I would expect this test to just output the first `blockquote`.

File: src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap [hunk 8]
```diff
@@ -0,0 +1,799 @@
...
+                >
+                  second list item
+                </span>
+              </span>
+            </div>
+          </li>
+        </ol>
+      </div>
+    </blockquote>
+    <div
```
Comment (flarnie) [inline src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: Similarly here - why is there another div with list items after the `blockquote`? And why is there an `ol` wrapping the `blockquote`?

File: src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap [hunk 9]
```diff
@@ -0,0 +1,799 @@
...
+      data-offset-key="C-0-0"
+    >
+      <span
+        data-offset-key="C-0-0"
+        style={Object {}}
+      >
+        <span
+          data-text="true"
+        >
+          deeply nested list
```
Comment (flarnie) [inline src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: This looks like what I would expect. :)

File: src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap [hunk 10]
```diff
@@ -0,0 +1,799 @@
...
+          data-offset-key="C-0-0"
+        >
+          <span
+            data-offset-key="C-0-0"
+            style={Object {}}
+          >
+            <span
+              data-text="true"
+            >
+              header inside list
```
Comment (flarnie) [inline src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: Again here the 'header inside list' item shows up twice in the output.

This is really interesting. The code and tests all make sense - I'll continue thinking about this and see if I can come up with theories about why the output isn't right.

**Comment by juliankrispel** [comment]
thanks @flarnie I'll look into it within the next two days (flying back to the UK tomorrow)

**Comment by flarnie** [comment]
Awesome - take your time and hope it is a good trip back! :)

File: src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap (hunk 7 above)
Comment (juliankrispel) [reply src/component/utils/exploration/__tests__/__snapshots__/wrapBlockNodes.js.snap]: @flarnie I think you mean two nested `ol` items, that's fixed now btw, about to upload the thing

**Comment by juliankrispel** [comment]
@flarnie the snapshots should all be good now! Please have another look :+1:

**Comment by flarnie** [comment]
Ok - going to look at this again, a bit today and probably again on the weekend. Starting by rebasing to see if we can get CI passing, and if there are any conflicts since Miter has continued changing/adding things.

**Comment by flarnie** [comment]
Also if it's convenient, it would be nice to move this to a branch on your fork of Draft @juliankrispel 
And thanks for your patience on the review! :)

**Comment by juliankrispel** [comment]
@flarnie <3 done - opened another branch >>> https://github.com/facebook/draft-js/pull/1608

**Comment by flarnie** [comment]
:D Awesome! Will look at this asap. Top of my Draft.js list.

**Comment by flarnie** [comment]
Thanks for opening another branch - closing this in favor of https://github.com/facebook/draft-js/pull/1608

**Pull request closed without merging**